### PR TITLE
POS Tagging widget

### PR DESCRIPTION
**Pull request by s-alexey**
@nikicc can you please describe what this widget is supposed to do?

We can tag `corpus.tokens` but where should we store results? And how would it cooperate with other widgets?

### CODE REVIEW

**Comment by nikicc** [comment]
@s-alexey we didn't really thought this through yet I'm afraid. 

One idea that immediately pops to mind is to enhance tokens with their POS tag. Hence, when doing classification one could distinguish between play used as a verb versus play used as a noun. This could be done by appending the tag after the token (e.g. `play_VB`). We could store this directly into `corpus.tokens`, though this is not very nice from the code design point of view, but it would instantly work for classification. 

Further POS tags could also be used for token filtering (but we would probably need an other widget for that). It might be interesting to compare different writing styles by observing how adjectives and verbs are used. For example, I could take a book, filter out everything but adjectives and observe this in a WordCloud.

**Comment by s-alexey** [comment]
Where should we take taggers? Currently, there are four taggers that can be obtained via `nltk`. 

And we can let users use Stanford POS taggers, but this will require them to provide `stanford-postagger.jar`.

**Comment by s-alexey** [comment]
One thing that we probably should bear in mind is that we may have different kind of tags (eg NER).

**Comment by nikicc** [comment]
> Where should we take taggers?

For a start we can work with the ones NLTK provides.

> We can also let users use Stanford POS taggers, but this will require them to provide stanford-postagger.jar.

We could allow this as an additional option, besides the taggers that will work out of the box.

**Comment by nikicc** [comment]
> One thing that we probably should bear in mind is that we may have different kind of tags (eg NER).

I agree. Though, I am not sure, if this should be a part of this widget.

**Comment by ajdapretnar** [comment]
I've tested this and I have some points I'd like to raise.

1.) The output is somewhat strange for me. I'm no expert, but shouldn't the user be able to see a BOW matrix with POS tags appended? I mean just like we have in BOW now, but with attribute name extended with POS tag? The problem is I see no transformation with Corpus viewer or Data Table. Only when I connect Word Cloud to it, I can see tags.

2.) Word Cloud now displays at the tags with the words. So a 'cat' is now 'cat_NN_NNP'. I'm not sure how to deal with this. It's a matter of discussion, but in my mind tags should be somehow hidden word attributes that one can access for modelling (tuples?). Maybe people wouldn't be connecting Word Cloud and POS Tagger anyway, so perhaps my concern is pointless. It's just that when I want to see tags, I don't, and when I don't, I do. :grinning: 

3.) Select Rows widget doesn't work with tags either. I imagine one might want to filter out only proper nous or adjectives for example.

4.) POS Tagger overrides the original data set, so when I connect Word Cloud or Topic Modelling to Corpus, all the words are tagged. Do we really want this or should we simply copy()?

5.) GUI is only slightly strange to me, but I cannot think of a good solution. I wouldn't have 'Stanford' repeat twice, once in radiobutton and once in box label. Also, while at it, Stanford doesn't work for me. When I set it, it cannot find the file with tagger. Shouldn't the file be downloaded with the widget?

That's it for now. :smiley:

**Comment by s-alexey** [comment]
> 4.) POS Tagger overrides the original data set, so when I connect Word Cloud or Topic Modelling to Corpus, all the words are tagged. Do we really want this or should we simply copy()?

We should copy 👍 Will change it.

File: orangecontrib/text/widgets/owpostagging.py [hunk 1]
```diff
@@ -93,7 +93,7 @@ def change_tagger(self, i):
             self.on_change()
 
     def set_data(self, data):
-        self.corpus = data
+        self.corpus = data.copy()
```
Comment (nikicc) [inline orangecontrib/text/widgets/owpostagging.py]: This will fail when `data` would be `None`.

Comment (s-alexey) [inline orangecontrib/text/widgets/owpostagging.py]: Yes, by bad.

**Comment by nikicc** [comment]
My thoughts, suggestions and discovered bugs:

- [ ] We should add a progress bar. 

- [ ] Concerning the design, maybe we could put all the radio buttons into a a box?

- [ ]  While loading Stanford JAR I get this error message `Problem while loading Stanford POS Tagger` along with this stack trace:
```
AssertionError                                Traceback (most recent call last):
  File "/Users/Niko/Projects/orange3-text/orangecontrib/text/widgets/owpostagging.py", line 126, in set_stanford_tagger
    StanfordPOSTagger.check(*paths)
  File "/Users/Niko/Projects/orange3-text/orangecontrib/text/tag/pos.py", line 27, in check
    cls(path_to_model, path_to_jar).tag(())
  File "/Users/Niko/Projects/virtual_envs/orange3env/lib/python3.4/site-packages/nltk/tag/stanford.py", line 136, in __init__
    super(StanfordPOSTagger, self).__init__(*args, **kwargs)
  File "/Users/Niko/Projects/virtual_envs/orange3env/lib/python3.4/site-packages/nltk/tag/stanford.py", line 56, in __init__
    env_vars=('STANFORD_MODELS',), verbose=verbose)
  File "/Users/Niko/Projects/virtual_envs/orange3env/lib/python3.4/site-packages/nltk/internals.py", line 573, in find_file
    file_names, url, verbose))
  File "/Users/Niko/Projects/virtual_envs/orange3env/lib/python3.4/site-packages/nltk/internals.py", line 466, in find_file_iter
    assert isinstance(filename, compat.string_types)
AssertionError
```

- [ ] If I click Report while Stanford tagger is selected but not loaded (there is none in the dropdown) I get this:
```
AttributeError                                Traceback (most recent call last):
  File "/Users/Niko/Projects/orange3/Orange/canvas/report/report.py", line 28, in show_report
    self.create_report_html()
  File "/Users/Niko/Projects/orange3/Orange/canvas/report/report.py", line 50, in create_report_html
    self.send_report()
  File "/Users/Niko/Projects/orange3-text/orangecontrib/text/widgets/owpostagging.py", line 134, in send_report
    self.report_items('Tagger', (('Name', self.tagger.name),))
AttributeError: 'NoneType' object has no attribute 'name'
```

- [ ] What is Trained Model button used for?

File: orangecontrib/text/widgets/owpostagging.py [hunk 2]
```diff
@@ -0,0 +1,143 @@
...
+        self.tagger = None
+        self.stanford_tagger = None
+
+        self.button_group = button_group = QtGui.QButtonGroup(self, exclusive=True)
+        button_group.buttonClicked[int].connect(self.change_tagger)
+
+        box = gui.vBox(self.controlArea, 'Method')
+        layout = QtGui.QGridLayout()
+        box.layout().addLayout(layout)
+        layout.setSpacing(25)
```
Comment (nikicc) [inline orangecontrib/text/widgets/owpostagging.py]: For my taste this spacing seems a bit too large. I think 10 would suffice.

**Comment by nikicc** [comment]
An other issue with Stanford POS Tagger:

- [ ] Loading either Tagger or Model triggers apply even when Stanford POS Tagger is not selected. IMHO these two buttons should trigger apply only when this tagger is selected.

**Comment by nikicc** [comment]
- [ ] Pos tags should also be copied in `Corpus.copy` method.

**Comment by nikicc** [comment]
- [ ] Now that we store pos tags as a separate attribute BoW should be adopted to consider them when present.

**Comment by nikicc** [comment]
- [ ] In reports for Stanford POS tagger we should also report the model's name.

**Comment by ajdapretnar** [comment]
>  Loading either Tagger or Model triggers apply even when Stanford POS Tagger is not selected. IMHO these two buttons should trigger apply only when this tagger is selected.

This goes somewhat against what we're trying to achieve in core Orange. The idea is, that is that if the user loads the files, she is doing it for the purpose of using Stanford tagger. For example, even if you don't select a method in core Orange, but you change its parameters, radio button would automatically be switched to the method whose parameters you're changing.

**Comment by nikicc** [comment]
> This goes somewhat against what we're trying to achieve in core Orange. The idea is, that is that if the user loads the files, she is doing it for the purpose of using Stanford tagger. For example, even if you don't select a method in core Orange, but you change its parameters, radio button would automatically be switched to the method whose parameters you're changing.

@ajdapretnar I agree with automatic switching of radio buttons. However, the process here requires two steps. I.e. you have to load both the Tagger and the Model in order to apply Stanford POS tagger. Loading only one does not suffice. Hence, I believe when you load the fist one nothing should happen. Once you load the second one it should automatically switch to Stanford and apply it. Do you agree?

**Comment by ajdapretnar** [comment]
:+1: Agree.

**Comment by nikicc** [comment]
@s-alexey any ideas what this is about?
```
Traceback (most recent call last):
  File "/Users/Niko/Projects/orange3-text/orangecontrib/text/widgets/owpostagging.py", line 143, in <module>
    widget = OWPOSTagger()
  File "/Users/Niko/Projects/orange3-text/orangecontrib/text/widgets/owpostagging.py", line 70, in __init__
    self.set_stanford_tagger(self.stanford.model_path, self.stanford.resource_path, silent=True)
  File "/Users/Niko/Projects/orange3-text/orangecontrib/text/widgets/owpostagging.py", line 123, in set_stanford_tagger
    self.button_group.buttonClicked.emit(self.STANFORD)
TypeError: QButtonGroup.buttonClicked[QAbstractButton].emit(): argument 1 has unexpected type 'int'
```

**Comment by s-alexey** [comment]
@nikicc it should be ok now

**Comment by nikicc** [comment]
@s-alexey it seem ok now. IMO two things still persist (one copied from above, one new):

- [ ]  In reports for Stanford POS tagger we should also report the model's name.

- [ ] The `Stanford POS Tagger` bullet should be enabled immediately after both are loaded and we trigger apply. Currently it is enabled only after the tagging has finished.

**Comment by nikicc** [comment]
And two more technical issues:

- [ ] Some commits are a duplicated from #98.
- [ ] Some additional test would be nice so coverage is happy.

**Comment by s-alexey** [comment]
> In reports for Stanford POS tagger we should also report the model's name.

Is should be the full path to model or just the filename?

**Comment by ajdapretnar** [comment]
Just file name IMO.

**Comment by ajdapretnar** [comment]
Huh. I cannot see the new attribute pos_tags in Corpus Viewer. Is this supposed to work already? Otherwise, everything works fine for me, with the obvious exception of Stanford tagger (my silly machine). One thing is that since my Stanford tagger is not working, I continue getting the error even when select other taggers instead. Also, when I place the POS tagger again on the canvas, java seems to run the console, even though my Stanford tagger is off. Is there any way to avoid this?

**Comment by ajdapretnar** [comment]
@nikicc Just to clarify how to tackle adding new attribute for Corpus Viewer. Did we agree on introducing a new **string** attribute which simply joins all tagged tokens? E.g.:
`pos_tags = "cat_NN, jumps_JJ, over_STH......."`
And then one can view all pos-tagged tokens for each document.

**Comment by nikicc** [comment]
> Just file name IMO.

I agree. We are using Stanford's pretrained models anyhow.

> Huh. I cannot see the new attribute pos_tags in Corpus Viewer. Is this supposed to work already? 

No, not yet 😞 

> @nikicc Just to clarify how to tackle adding new attribute for Corpus Viewer. Did we agree on introducing a new string attribute which simply joins all tagged tokens? E.g.:
pos_tags = "cat_NN, jumps_JJ, over_STH......."
And then one can view all pos-tagged tokens for each document.

Not sure if we agreed on this yet. We should support viewing all possibilities IMO. I.e. tokens, n-grams and pos-tags. And, since these are not features in the domain maybe we should do something else than just introduce them as string attributes.

File: doc/widgets/postagging.rst [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+**Outputs**:
+
+-  **Corpus**
+
+Marked corpus.
+
+Description
+-----------
+
+This widgets adds POS tags to each token in documents. This tags could be used in BagOfWords widget to distinguish
```
Comment (nikicc) [inline doc/widgets/postagging.rst]: This widget~~s~~ adds ...

Comment (nikicc) [inline doc/widgets/postagging.rst]: Also: ~~This~~ These tags ....

File: doc/widgets/postagging.rst [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+tokens with the same spelling.
+
+Stanford POS Tagger
+^^^^^^^^^^^^^^^^^^^
+
+This widgets also allows you to load `Stanford POS Taggers <http://nlp.stanford.edu/software/tagger.shtml>`_.
```
Comment (nikicc) [inline doc/widgets/postagging.rst]: widgets -> widget

File: doc/widgets/postagging.rst [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+Download tagger you interested in and provide path to model along with ``stanford-postagger.jar``.
```
Comment (nikicc) [inline doc/widgets/postagging.rst]: you _are_ interested in.

File: orangecontrib/text/corpus.py [hunk 6]
```diff
@@ -238,15 +239,24 @@ def from_file(cls, filename):
...
         if join_with is None:
-            return (list(chain(*(nltk.ngrams(doc, n)
-                    for n in range(self.ngram_range[0], self.ngram_range[1]+1))))
-                    for doc in self.tokens)
+            processor = lambda doc, n: nltk.ngrams(doc, n)
         else:
-            return (list(chain(*((join_with.join(ngram) for ngram in nltk.ngrams(doc, n))
-                    for n in range(self.ngram_range[0], self.ngram_range[1]+1))))
-                    for doc in self.tokens)
+            if include_postags:
```
Comment (nikicc) [inline orangecontrib/text/corpus.py]: Instead of two nested if-else blocks could this just be an if-elif-else block?

Comment (s-alexey) [inline orangecontrib/text/corpus.py]: Yes, it could

File: orangecontrib/text/tag/pos.py [hunk 7]
```diff
@@ -0,0 +1,67 @@
...
+
+    def tag_corpus(self, corpus, **kwargs):
+        """ Marks tokens of a corpus with POS tags.
+
+        Args:
+            corpus (orangecontrib.text.corpus.Corpus): A corpus instance.
+
+        """
+        corpus.pos_tags = np.array(self._tag_sents(corpus.tokens, **kwargs), dtype=object)
+        # corpus.store_tokens(list(map(lambda sent: list(map(lambda x: '{0[0]}_{0[1]}'.format(x), sent)), tags)))
```
Comment (nikicc) [inline orangecontrib/text/tag/pos.py]: We could delete this now, right?

File: orangecontrib/text/topics/topics.py [hunk 8]
```diff
@@ -70,14 +60,9 @@ def reset_model(self, corpus):
...
-    def update(self, corpus, progress_callback=None):
-        chunk_size = np.ceil(len(corpus) / 100)
-        for i, chunk in enumerate(chunks(corpus.ngrams_corpus, chunk_size=chunk_size)):
-            if not self.running:
-                break
-            self.model.update(chunk)
-            if progress_callback:
-                progress_callback(100 * (i + 1) * chunk_size / len(corpus))
+    @chunkable
+    def update(self, documents):
```
Comment (nikicc) [inline orangecontrib/text/topics/topics.py]: This is really nice 👍

File: orangecontrib/text/tag/pos.py (hunk 7 above)
Comment (s-alexey) [inline orangecontrib/text/tag/pos.py]: Yes, probably

File: orangecontrib/text/widgets/utils/owwidget.py [hunk 9]
```diff
@@ -0,0 +1,88 @@
...
+from PyQt4.QtCore import pyqtSlot as Slot, QMetaObject, Qt, Q_ARG
+
+from Orange.widgets.widget import OWWidget
+
+
+class StopExecution(Exception):
+    pass
+
+
+class OWConcurrentWidget(OWWidget):
```
Comment (nikicc) [inline orangecontrib/text/widgets/utils/owwidget.py]: I thought that `@chunkable` was nice, but this ... this is the real elegance 😉 👍 👏

**Comment by nikicc** [comment]
Looks good to me, so I'm merging. I especially like the `OWConcurrentWidget`.

File: orangecontrib/text/widgets/utils/owwidget.py (hunk 9 above)
Comment (kernc) [inline orangecontrib/text/widgets/utils/owwidget.py]: This is something we could use in [biolab/orange3](https://github.com/biolab/orange3)!

Comment (ajdapretnar) [inline orangecontrib/text/widgets/utils/owwidget.py]: I am telling you, Alexey has really great ideas! :smiley:

**Pull request merged**